src/kixi/nhs/board_report.clj

```diff
@@ -23,35 +24,145 @@
 
 (defn not-nil? [x] (not (nil? x)))
 
+(defn all-fields-exist? [fields row]
+  (let [headers (into #{} (keys row))]
+    (every? #(contains? headers %) fields)))
+
 (defn filter-dataset
   "Filters dataset according to the given recipe."
   [recipe-map data]
-  (let [{:keys [indicator-field conditions indicator-id]} recipe-map]
-    ;; Go through the data sequence and 1.Check indicator field,
-    ;; 2.Check the conditions, 3.Keep "Year" and indicator value.
-    (when (contains? (first data) indicator-field)
+  (let [{:keys [conditions indicator-id fields-to-extract]} recipe-map]
+    (when (all-fields-exist? fields-to-extract (first data))
       (keep (fn [d] (when (every? (fn [condition] (let [{:keys [field values]} condition]
                                                     ;; values is a set
                                                     (some values #{(get d field)})))
                                   conditions)
-                      (select-keys d [:year :period_of_coverage indicator-field])))
+                      (select-keys d fields-to-extract)))
             data))))
 
 (defn enrich-dataset
   "Enrichs dataset with indicator-id."
   [recipe-map data]
-  (let [{:keys [indicator-field indicator-id]} recipe-map]
-    ;; Go through the data sequence and 1.Change the key map indicator-field
-    ;; for value, 2.Add the indicator-id.
+  (let [{:keys [indicator-id]} recipe-map]
     (mapv (fn [d]
             (-> d
-                (clojure.set/rename-keys {indicator-field :value})
                 ;; period_of_coverage is a PK so cannot be null. Using year if it's empty
                 (cond-> (empty? (:period_of_coverage d)) (assoc :period_of_coverage (:year d)))
                 (assoc :indicator_id indicator-id))) data)))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;; Internal calculations                                                                ;;
+;; Internal calculations - indicator: 212                                               ;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn split-by-key
+  "Turns a sequence of maps into a sequence of sequences,
+  where each nested sequence corresponds to a single group."
+  [k data]
+  (->> (group-by k data)
+       vals))
+
+(defn get-slope [y]
+  (let [{:keys [year period_of_coverage]} (first y)
+        x [0.1 0.3 0.5 0.7 0.9]
```

Comment:
Happy to have them in the code. I just want to know why they are there as they appeaer to be magic numbers.